fleetspeak/src/client/channel/channel.go

```diff
@@ -211,34 +264,60 @@ func (c *Channel) writeLoop() {
 		c.e <- fmt.Errorf("error writing magic number: %v", err)
 		return
 	}
-	if !<-c.magicRead {
-		return
-	}
-	for msg := range c.o {
-		buf, err := proto.Marshal(msg)
+	// Send startup data.
+	if c.startupDataOut != nil {
+		sd, err := ptypes.MarshalAny(c.startupDataOut)
 		if err != nil {
-			log.Errorf("Unable to marshal message to send: %v", err)
-			continue
-		}
-		if len(buf) > int(maxSize) {
-			log.Errorf("Overlarge message to send, want less than [%x] got [%x]", maxSize, len(buf))
-			continue
-		}
-		if err := binary.Write(c.pipeWrite, byteOrder, int32(len(buf))); err != nil {
-			c.e <- fmt.Errorf("error writing message length: %v", err)
+			c.e <- fmt.Errorf("unable to marshal StartupData: %v", err)
 			return
 		}
-		if _, err := c.pipeWrite.Write(buf); err != nil {
-			c.e <- fmt.Errorf("error writing message: %v", err)
+		// We do not set a destination for the initial message, since it is not meant for delivery
+		// to any particular component (other than the channel implementation).
```

Comment:
In the RelentlessChannel protocol, we identify local control messages by setting m.Source.ServiceName to "client" which is a reserved service name (FS won't let you define a service named 'client' or 'system', and the system name is used for all FS internal messages between the client and server). We might want to match this, mostly for consistency.

https://github.com/google/fleetspeak/blob/da9e4855ff43948fea8ced41020695285e51f2e2/fleetspeak/src/client/channel/relentless.go#L173